src/main/java/io/gingersnapproject/cdc/ManagedEngine.java

```diff
@@ -28,19 +30,21 @@ public class ManagedEngine implements DynamicRuleManagement {
    private static final Logger log = LoggerFactory.getLogger(ManagedEngine.class);
    private static final ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(r ->
          new Thread(r, "scheduled-engine-error-handler"));
-   private final Map<String, StartStopEngine> engines = new ConcurrentHashMap<>();
+   private final Map<CacheIdentifier, StartStopEngine> engines = new ConcurrentHashMap<>();
+   private final Map<String, Rule> knownRules = new ConcurrentHashMap<>();
+   private final Set<URI> knownMembers = ConcurrentHashMap.newKeySet();
+   private final Set<URI> offlineMembers = ConcurrentHashMap.newKeySet();
```

Comment:
This set could grow indefinitely if the URIs keep changing, but this is also an issue with the `HotRodService`, which keeps the values to identify restarts. Another approach is to put together the start and reconnection parts in the `HotRodService` to be invisible. We use the scheduled task to start a cache.